### ISSUE TITLE
R CMD check fails with dev roxygen2

### ISSUE THREAD

**Issue by hadley** [torfason/zmisc#1]
```
❯ checking re-building of vignette outputs ... WARNING
  Error(s) in re-building vignettes:
    ...
  --- re-building ‘zmisc.Rmd’ using rmarkdown
  Quitting from lines 57-69 (zmisc.Rmd) 
  Error: processing vignette 'zmisc.Rmd' failed with diagnostics:
  cannot unclass an environment
  --- failed re-building ‘zmisc.Rmd’
  
  SUMMARY: processing the following file failed:
    ‘zmisc.Rmd’
  
  Error: Vignette re-building failed.
  Execution halted
```

Not sure what's going wrong here, but if you can make a simple reprex, I'm happy to take a look. I'm aiming for a CRAN release May 13.

**Comment by torfason** [comment]
Thanks for the report. I think I've figured out what the issue is.

The error is occurring inside a function I cobbled together based on internal `roxygen2` source code, and which calls private roxygen functions left and right. The reason I put together this function is that I wanted to be able to programmatically extract unparsed markdown versions of my `roxydoc` documentation fields (`@title`, `@description`, `@examples`) for use in vignettes. For this, I call `blocks <- roxygen2:::parse_package(base_path, env = NULL)` to get a roxygen documentation block structure. 

The error then occurs when I call `rapply(blocks, unclass)` because when using the dev version, there is an item in `blocks` that is of mode `environment`, that wasn't there before (when using the release version).

The fix is straight enough, I can remove the unclassing, it is not needed.  

A much nicer fix, though, would be to get rid of the two functions I wrote, and I wonder if you know of another way to do what I'm trying to do, without relying on internal code?

I would assume the immediate concern is for your own submission, and to address reverse dependency notes from CRAN. I don't have specific expertise on this, but I would think that just blaming me for indiscriminately calling internal code would suffice :-)  

I'll then try to roll a CRAN release sooner rather than later, but I'm thinking that this will not cause issues with any actual users, because vignettes are not rebuilt on install. Or am I wrong?

The reprex, including the two functions follows:

``` r
# Parse source code to get roxygen block objects
#
# This function is based on roxygenise, but omits the parts that have side
# effects - that is the parts that affect the enviroment and that output
# documentation files.
roxy_get_blocks <- function (package.dir = ".") {
  load_code <- NULL
  clean <- FALSE
  base_path <- normalizePath(package.dir, mustWork = TRUE)
  encoding <- desc::desc_get("Encoding", file = base_path)[[1]]
  if (!identical(encoding, "UTF-8")) {
      warning("roxygen2 requires Encoding: UTF-8", call. = FALSE)
  }
  roxygen2:::roxy_meta_load(base_path)
  packages <- roxygen2:::roxy_meta_get("packages")
  lapply(packages, loadNamespace)
  load_code <- roxygen2:::find_load_strategy(load_code)
  env <- load_code(base_path)
  blocks <- roxygen2:::parse_package(base_path, env = NULL)
  blocks <- lapply(blocks, roxygen2:::block_set_env, env = env)
  blocks
}


# Extract a documentation section from roxy blocks object
#
# This function takes a roxygen blocks structure, which typically is created by
# calling roxy_get_blocks(), and extracts the content for a particular topic
# (usually a function name) and tag (such as @title/@description/@examples).
#
# The main use case is to include the result in custom documentation files. The
# result is returned in the form of a character vector, which contains any
# formating as it is written in the source file. Thus, if the intention is to
# use the result in an .Rmd file, the roxydoc documentation should be formatted
# using markdown (and the @md tag should therefore be included)
#
# @param blocks A roxy blocks structure (list)
# @param name_topic The name of the topic (usually a function name)
# @param name_tag The name of the tag/section (@title/@description/@examples)
roxy_get_section <- function(blocks, name_topic, name_tag) {

  # Better to work with unclassed structure
  b <- blocks |> rapply(unclass, how="list")

  # Get the list of all topics (function names) that are
  # documented in the package, and assign as names to b
  topics <- b |>
    lapply('[[', 'object') |>
    lapply('[[', 'topic') |>
    unlist()
  names(b) <- topics

  # Get the substructure for the current topic, extract
  # all its tags, and assign as names to the structure
  b.cur_topic <- b[[name_topic]]$tags
  cur_topic_tags <- b.cur_topic |>
    lapply('[[', 'tag') |>
    unlist()
  names(b.cur_topic) <- cur_topic_tags

  # What we are interested in is the 'raw' field in
  # of the focal tag of the topic
  result <- b.cur_topic[[name_tag]]$raw

  # Return the result, trimming whitespace at start/end
  stringr::str_trim(result)
}

blocks <- roxy_get_blocks(rprojroot::find_package_root_file())
#> ℹ Loading zmisc
blocks <- force(blocks)
roxy_get_section(blocks, "lookup", "title") |> cat()
#> Error in .Primitive("unclass")(X, ...): cannot unclass an environment
roxy_get_section(blocks, "lookup", "description") |> cat()
#> Error in .Primitive("unclass")(X, ...): cannot unclass an environment
```

<sup>Created on 2022-04-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

And the tail end, after a fix by changing the `b <- blocks |> rapply(unclass, how="list")` line to `b <- blocks # |> rapply(unclass, how="list")` becomes like this:

``` r
blocks <- roxy_get_blocks(rprojroot::find_package_root_file())
#> ℹ Loading zmisc
blocks <- force(blocks)
roxy_get_section(blocks, "lookup", "title") |> cat()
#> Lookup values from a lookup table
roxy_get_section(blocks, "lookup", "description") |> cat()
#> The [lookup()] function implements lookup of certain strings (such as
#> variable names) from an lookup table which maps keys onto values (such as
#> variable labels or descriptions). Original values are returned if they are
#> not found in the lookup table.
#> 
#> The lookup table can be in the form of a two-column `data.frame`, in the form
#> of a named `vector`, or in the form of a `list`. If the table is in the form
#> of a `data.frame`, the lookup columns should be named `name` (for the key)
#> and `value` (for the value). If the lookup table is in the form of a named
#> `vector` or `list`, the name is used for the key, and the returned value is
#> taken from the values in the vector or list.
#> 
#> Any names in x are not included in the result.
```

**Comment by hadley** [comment]
Could you do it the other way round? i.e. write in Rmd and then include that into your docs with `@includeRmd`?

(In terms of CRAN submission, this issue is enough to demonstrate that I've done my due diligence informing you of the problem and giving you time to respond. If you don't fix before release, you'll get an automated notification from CRAN and giving you ~4 weeks to fix, so there's still plenty of time)e

**Comment by torfason** [comment]
I actually started `@includeRmd` and it works great, but the drawback for my application was that:

- It means the docs are further from the source (sepearate `.Rmd` file instead of the `.R` file)
- I could not get it to work with examples, whereas `roxy_get_section(blocks, "zample", "examples")` gives a string containing the `zample()` function examples, which can be embedded in a markdown file and run there.

But I digress :-)

The upshot is that since the immediate fix (removing `unclass()` seemed straightforward I've rolled a small release with the fix (`v0.2.2`) and pushed it to CRAN, where everything seems to have gone smoothly. So I'm closing, hopefully this will fix the `revdep` issues – if not, please reopen the issue.

**Comment by hadley** [comment]
Thanks!
